app/services/bulk_update/add_trainees/import_rows.rb

```diff
@@ -0,0 +1,128 @@
+        "ITT Aim" => "itt_aim",
+        "Training Route" => "training_route",
+        "Qualification Aim" => "itt_qualification_aim",
+        "Course Subject One" => "course_subject_one",
+        "Course Subject Two" => "course_subject_two",
+        "Course Subject Three" => "course_subject_three",
+        "Study Mode" => "study_mode",
+        "ITT Start Date" => "itt_start_date",
+        "ITT End Date" => "itt_end_date",
+        "Course Age Range" => "course_age_range",
+        "Course Year" => "course_year",
+        "Lead Partner URN" => "lead_partner_urn",
+        "Employing School URN" => "employing_school_urn",
+        "Trainee Start Date" => "trainee_start_date",
+        "PG Apprenticeship" => "pg_apprenticeship",
+        "Start Date" => "start_date",
+        "Fund Code" => "fund_code",
+        "Funding Method" => "funding_method",
+        "Training Initiative" => "training_initiative",
+        "Additional Training Initiative" => "additional_training_initiative",
+      }.freeze
+
+      PLACEMENT_HEADERS = {
+        "First Placement URN" => "urn",
+      }.freeze
+
+      DEGREE_HEADERS = {
+        "UK Degree Type" => "uk_degree_type",
+        "Non-UK Degree Type" => "non_uk_degree_type",
+        "Degree Subject" => "degree_subject",
+        "Degree Grade" => "degree_grade",
+        "Degree Graduation Year" => "degree_graduation_year",
+        "Degree UK" => "degree_uk",
+        "Awarding Institution" => "degree_uk",
+        "Degree Country" => "degree_country",
+      }.freeze
+
+      def call
+        return unless FeatureService.enabled?(:bulk_add_trainees)
+
+        dry_run = !trainee_upload.submitted?
+        success = true
+
+        ActiveRecord::Base.transaction do |_transaction|
+          if dry_run
+            CSV.parse(trainee_upload.file, headers: true).map.with_index do |row, index|
+              BulkUpdate::TraineeUploadRow.create!(
+                bulk_update_trainee_upload: trainee_upload,
+                data: row.to_h,
+                row_number: index + 1,
+              )
+            end
+          end
+
+          ActiveRecord::Base.transaction do |_transaction|
+            results = trainee_upload.bulk_update_trainee_upload_rows.map do |upload_row|
+              BulkUpdate::AddTrainees::ImportRow.call(row: upload_row.data, current_provider: current_provider)
+            end
+
+            # Commit or rollback the transaction depending on whether all rows were error free
+            if all_succeeded?(results)
+              trainee_upload.succeeded! unless dry_run
+            else
+              # TODO: copy any errors into `trainee_upload`
+              success = false
+              raise(ActiveRecord::Rollback)
+            end
+            raise(ActiveRecord::Rollback) if dry_run
+          end
+
+          trainee_upload.validated! if dry_run && success
+          trainee_upload.failed! unless success
+        end
+
+        success
+      end
```

Comment:
If any of the ActiveRecord objects raises an exception, I think would be good to mark the `trainee_upload` as `failed`.

ex:

```
rescue ActiveRecord::ActiveRecordError
  trainee_upload.failed!

  raise 
end